{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# transcriptResolver\n",
    "### Reconciling multiple transcripts using sequence alignment algorithms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You will require a number of python packages to run this. You can do this by using the pip installer. If you're running Python 3, this involves inputing the following in Terminal\n",
    "\n",
    "<code> pip3 install fuzzywuzzy numpy pandas nltk python-Levenshtein </code>\n",
    "\n",
    "To run transcriptResolver in Terminal,\n",
    "\n",
    "<code> python3 transcriptResolver.py </code>\n",
    "\n",
    "You will also have to install mafft\n",
    "\n",
    "<code> brew install mafft </code>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%file transcriptResolver.py\n",
    "\n",
    "import csv\n",
    "import os\n",
    "import itertools as it\n",
    "import numpy as np\n",
    "import re # regular expressions\n",
    "import argparse #For command line arguments\n",
    "\n",
    "from consensus_tools import * # custom functions to run transcript resolving\n",
    "from collections import defaultdict # utility functions to create dictionaries\n",
    "import pandas as pd # data frame functionality\n",
    "from fuzzywuzzy import process, fuzz # Functions that are useful for fuzzy string matching (https://github.com/seatgeek/fuzzywuzzy)\n",
    "from functools import reduce # for the reduce function\n",
    "\n",
    "import webbrowser\n",
    "import copy\n",
    "\n",
    "class transcriptResolver:\n",
    "    def __init__(self, args): # __init__ always run when an instance of the class is created\n",
    "\n",
    "        ## Define stem name ========================\n",
    "        if args.stem:\n",
    "            self.stem = args.stem\n",
    "        else:\n",
    "            print(\"\\nPlease input a 'stem' name to act as a prefix to all output (e.g., stemname_calbug.csv)\")\n",
    "            self.stem = input(\"Stem name: \")\n",
    "        \n",
    "        print(\"\\nStem name for all outputs will be '\" + self.stem + \"'\")\n",
    "        \n",
    "        ## Define working directory ========================\n",
    "        if args.wd:\n",
    "            self.wd = args.wd\n",
    "        else:\n",
    "            temp = input(\"Working directory: \")\n",
    "            self.wd = temp\n",
    "\n",
    "        print(\"\\nUsing working directory '\" + self.wd + \"' ...\")\n",
    "\n",
    "        ## Define transcription file ========================\n",
    "        if args.file:\n",
    "            tempfile = args.file\n",
    "        else:\n",
    "            tempfile = input(\"\\nInput your working file name. \\nWorking file should be in your stated working directory:\")\n",
    "        \n",
    "        filedir = os.path.join(self.wd, tempfile)\n",
    "        print(\"\\nFile directory will be '\" + filedir + \"'\")\n",
    "\n",
    "        \n",
    "        ## Define id column ========================\n",
    "        ##todo## need to check if ID column is in the file\n",
    "        if args.col_id:\n",
    "            self.col_id = args.col_id\n",
    "        else:\n",
    "            self.col_id = input(\"\\nDefine the column name specifying unique IDs (e.g., specimen ID):\")\n",
    "\n",
    "        print(\"\\nColumn name that specifies unique ID is \" + self.col_id)\n",
    "        \n",
    "        ## Define target columns ========================\n",
    "        ##todo## need to check if target columns are in the file\n",
    "        targetlist = []\n",
    "        methodlist = []\n",
    "        \n",
    "        if args.col_target and args.col_method:\n",
    "            temp_target = args.col_target.strip(\"[|]\").split(\",\")\n",
    "            temp_method = args.col_method.strip(\"[|]\").split(\",\")\n",
    "            if len(temp_target) > 0 and len(temp_method) > 0 and len(temp_method) == len(temp_target):\n",
    "                self.col_target = temp_target\n",
    "                self.col_method = temp_method\n",
    "        \n",
    "        else:\n",
    "            while True:\n",
    "                temp_target = input(\"\\nDefine the column name to be resolved:\\n (Enter nothing to continue to the next step)\")\n",
    "                temp_method = input(\"\\nPlease define the method for which you would like to use on this column: \\n(Enter nothing to continue to the next step)\")\n",
    "            \n",
    "                if(temp_method == \"\" or temp_target == \"\"):\n",
    "                    break            \n",
    "                \n",
    "                targetlist.append(temp_target)\n",
    "                methodlist.append(temp_target)\n",
    "            \n",
    "            self.col_target = targetlist\n",
    "            self.col_method = methodlist\n",
    "                \n",
    "        [print(\"\\nUsing method\", y, \"for column\", x) for x, y in zip(self.col_target, self.col_method)]\n",
    "        \n",
    "        ## Import file ========================\n",
    "        allcols = copy.copy(self.col_target) # make a copy so we don't alter self.col_target\n",
    "        allcols.append(self.col_id)\n",
    "        \n",
    "        self.file = pd.read_csv(filedir, dtype=object,\\\n",
    "                                encoding = \"ISO-8859-1\",\n",
    "                                usecols = allcols) # only use columns that were supplied\n",
    "        self.file = self.file.fillna(\"\") # Converts all NaNs into empty strings for alignment\n",
    "        \n",
    "        \n",
    "## MAIN ##\n",
    "def main():\n",
    "    args = parser.parse_args()\n",
    "    if args.version:\n",
    "        print(\"v1.0\")\n",
    "    elif args.manual:\n",
    "        webbrowser.open(\"https://github.com/junyinglim/TranscriptResolver\")\n",
    "    else:\n",
    "        print(\"\\n\\n\\n\")\n",
    "        print(\"=\" * 50)\n",
    "        print(\"WELCOME TO TRANSCRIPT RESOLVER!!\")\n",
    "        print(\"Let's resolve some replicate transcripts! \\n\")\n",
    "        print(\"Please visit https://github.com/junyinglim/Notes-from-Nature \\nfor a short explanation of the transcript resolution methods available \\n\")\n",
    "        print(\"This crude program was written by Jun Ying Lim (junyinglim@gmail.com) \\nfor the Essig Museum of Entomology at UC Berkeley\")\n",
    "        print(\"=\" * 50)\n",
    "        print(\"\\n\\n\\n\")\n",
    "               \n",
    "    ## Startup\n",
    "    currentArgs = transcriptResolver(args)\n",
    "    \n",
    "    # Create empty list\n",
    "    results = []\n",
    "    for col_no in range(len(currentArgs.col_target)):\n",
    "        if currentArgs.col_method[col_no] == \"vote_count\":\n",
    "            df = vote_count(accession = currentArgs.col_id,\\\n",
    "                            field = currentArgs.col_target[col_no],\\\n",
    "                            data = currentArgs.file)\n",
    "        elif currentArgs.col_method[col_no] == \"consensus\":\n",
    "            df = variant_consensus(accession = currentArgs.col_id,\\\n",
    "                                   field = currentArgs.col_target[col_no],\\\n",
    "                                   align_method = \"character\",\\\n",
    "                                   consensus_method = \"dumber\",\\\n",
    "                                   wdir = currentArgs.wd,\\\n",
    "                                   data = currentArgs.file)\n",
    "        elif currentArgs.col_method[col_no] == \"metadata\":\n",
    "            df = metadata_handling(accession = currentArgs.col_id,\\\n",
    "                                   field = currentArgs.col_target[col_no],\\\n",
    "                                   data = currentArgs.file)\n",
    "        else:\n",
    "            ##todo## write a proper error handling here\n",
    "            print(\"Sorry, method supplied is not valid\")\n",
    "        \n",
    "        # Add data frame to the results list\n",
    "        results.append(df)\n",
    "        \n",
    "    # Merge results\n",
    "    allResults = reduce(lambda a, d: pd.merge(a, d, on = currentArgs.col_id), results)\n",
    "            \n",
    "    finalDir = os.path.join(currentArgs.wd, currentArgs.stem + \"_results.csv\")\n",
    "    allResults.to_csv(finalDir, index = False)\n",
    "    print(\"\\nExporting results to\", finalDir)\n",
    "    \n",
    "\n",
    "if __name__ == '__main__':\n",
    "    parser = argparse.ArgumentParser(description=\"phyloGenerator - phylogeny generation for ecologists.\", epilog=\"Help at http://willpearse.github.com/phyloGenerator - written by Will Pearse\")\n",
    "    parser.add_argument(\"--version\", action=\"store_true\", help=\"Display version information.\")\n",
    "    parser.add_argument(\"--manual\", action=\"store_true\", help=\"(Attempt to) open browser and show help\")\n",
    "    parser.add_argument(\"-stem\", \"-n\", help=\"'Stem' name for all output files.\") # for command line\n",
    "    parser.add_argument(\"-wd\", help = \"Working directory\")\n",
    "    parser.add_argument(\"-file\", \"-f\", help = \"File with transcriptions\")\n",
    "    parser.add_argument(\"-col_id\", help = \"List of columns to be resolved\")\n",
    "    parser.add_argument(\"-col_target\", help = \"Target column. Must be in the format -col_target [target1,target2,target3]\")\n",
    "    parser.add_argument(\"-col_method\", help = \"Method. Must be in the format -col_method [method1,method2,method3]\")\n",
    "    main()\n",
    "    \n",
    "##todo## logging the results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running transcriptResolver\n",
    "You can either input the arguments in a string, or input them when prompted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "==================================================\n",
      "WELCOME TO TRANSCRIPT RESOLVER!!\n",
      "Let's resolve some replicate transcripts! \n",
      "\n",
      "Please visit https://github.com/junyinglim/Notes-from-Nature \n",
      "for a short explanation of the transcript resolution methods available \n",
      "\n",
      "This crude program was written by Jun Ying Lim (junyinglim@gmail.com) \n",
      "for the Essig Museum of Entomology at UC Berkeley\n",
      "==================================================\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Stem name for all outputs will be 'BLAH'\n",
      "\n",
      "Using working directory '/Users/junyinglim/Dropbox/Projects/2013/Calbug/TranscriptResolver/tests' ...\n",
      "\n",
      "File directory will be '/Users/junyinglim/Dropbox/Projects/2013/Calbug/TranscriptResolver/tests/test_file.csv'\n",
      "\n",
      "Column name that specifies unique ID is UNIQUE_ID\n",
      "\n",
      "Using method vote_count for column DATE\n",
      "\n",
      "Using method vote_count for column TARGET\n",
      "Implementing vote-counting procedure on DATE field.\n",
      "Reconciling transcriptions for A\n",
      "Reconciling transcriptions for B\n",
      "Implementing vote-counting procedure on TARGET field.\n",
      "Reconciling transcriptions for A\n",
      "Reconciling transcriptions for B\n",
      "\n",
      "Exporting results to /Users/junyinglim/Dropbox/Projects/2013/Calbug/TranscriptResolver/tests/BLAH_results.csv\n"
     ]
    }
   ],
   "source": [
    "%run transcriptResolver -wd ~/Dropbox/Projects/2013/Calbug/TranscriptResolver/tests \\\n",
    "                        -file test_file.csv \\\n",
    "                        -stem BLAH \\\n",
    "                        -col_id UNIQUE_ID\\\n",
    "                        -col_target [DATE,TARGET] \\\n",
    "                        -col_method [vote_count,vote_count]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "Or alternatively, you can run a prompt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%run transcriptResolver.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
